test_conformance/geometrics/test_geometrics_fp.cpp

```diff
@@ -0,0 +1,2240 @@
+        clCreateBuffer(context, CL_MEM_READ_WRITE, dstBufSize, NULL, &error);
+    if (streams[2] == NULL)
+    {
+        print_error(error, "ERROR: Creating output array failed!\n");
+        return -1;
+    }
+
+    /* Assign streams and execute */
+    for (i = 0; i < 3; i++)
+    {
+        error = clSetKernelArg(kernel, (int)i, sizeof(streams[i]), &streams[i]);
+        test_error(error, "Unable to set indexed kernel arguments");
+    }
+
+    /* Run the kernel */
+    threads[0] = test_size;
+    error = get_max_common_work_group_size(context, kernel, threads[0],
+                                           &localThreads[0]);
+    test_error(error, "Unable to get work group size to use");
+
+    error = clEnqueueNDRangeKernel(queue, kernel, 1, NULL, threads,
+                                   localThreads, 0, NULL, NULL);
+    test_error(error, "Unable to execute test kernel");
+
+    /* Now get the results */
+    error = clEnqueueReadBuffer(queue, streams[2], true, 0, dstBufSize, outData,
+                                0, NULL, NULL);
+    test_error(error, "Unable to read output array!");
+
+    /* And verify! */
+    int skipCount = 0;
+    for (i = 0; i < test_size; i++)
+    {
+        T *src1 = inDataA + i * vecSize;
+        T *src2 = inDataB + i * vecSize;
+        double expected = p.verifyFunc(src1, src2, vecSize);
+        bool isDif = (std::is_same<T, half>::value)
+            ? (HFF(expected) != outData[i])
+            : ((T)expected != outData[i]);
+        if (isDif)
+        {
+            if (std::is_same<T, half>::value)
+            {
+                T expv = HFF(expected);
+                if (isnan_fp<T>(expv) && isnan_fp<T>(outData[i])
+                    || (!isfinite_fp<T>(expv) && !isfinite_fp<T>(outData[i])))
+                    continue;
+            }
+            else if (isnan(expected) && isnan_fp<T>(outData[i]))
+                continue;
+
+            if (std::is_same<T, float>::value && !floatHasInfNan
+                || std::is_same<T, half>::value && !halfHasInfNan)
+            {
+                for (size_t ii = 0; ii < vecSize; ii++)
+                {
+                    if (!isfinite_fp<T>(src1[ii]) || !isfinite_fp<T>(src2[ii]))
+                    {
+                        skipCount++;
+                        continue;
+                    }
+                }
+                if (!isfinite_fp(expected))
+                {
+                    skipCount++;
+                    continue;
+                }
+            }
+
+            std::stringstream sstr;
+            std::string printout;
+            if (ulpLimit < 0)
+            {
+                // Limit below zero means we need to test via a computed error
+                // (like cross product does)
+                T maxValue = GetMaxValue<T>(inDataA + i * vecSize,
+                                            inDataB + i * vecSize, vecSize, p);
+
+                double error = 0.0, errorTolerance = 0.0;
```

Comment:
I'm seeing errors for cross and dot tests. The output from the verification error itself is informative, so nice work there, but as an aside the `CL_DEVICE_NOT_FOUND` in the backtrace is slightly confusion.

cross fail
```
ERROR: Data sample 122 does not validate! Expected (inf,inf,-0x1.b28p+15,0x0p+0), got (0x0p+0,0x1.0040140100001p+2,0x0p+0,0x1.0040140100401p+2)
    Input: (0x1.cbp+7 -0x1.9d8p+6 0x1.9dcp+8) and (0x1.824p+8 -0x1.a04p+8 -0x1.2dcp+6)
    Errors: (nan out of 0x1.fb9c0cp+8), (nan out of 0x1.f587ccp+8), (inf out of 0x1.fb9c0cp+8)
     ulp 0.000000
ERROR: CrossFPTest::RunSingleTest: test_relational failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.cpp:1012)
ERROR: GeometricsFPTest::Run: test_relational failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.cpp:933)
ERROR: Test Failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.h:325)
geom_cross FAILED
```

dot failed
```
ERROR: Data sample 122 at size 2 does not validate! Expected (-0x1.03b78p+14), got (-inf), sources (0x1.8dp+8 and -0x1.7acp+7) error of inf against tolerance 499.109
        vector A: { 397, 141.875,}, vector B: { -189.375, 412.75,}   geom_two_args vector size 2 FAILED
ERROR: TwoArgsFPTest::RunSingleTest: test_geometrics failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.cpp:1175)
ERROR: GeometricsFPTest::Run: test_relational failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.cpp:933)
ERROR: Test Failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.h:325)
geom_dot FAILED
```

From debugging this and why our internal testing didn't pick it up, I interpreted the spec absolute error calculations as being done in half precision. This meant that `'max * max * (2n - 1) * HLF_EPSILON'`
could overflow in an intermediate calculation. Essentially giving us
more leeway for our dot implementation to overflow.

I think we should have a spec clarification for all of half, float, and double.
dot/crosss absolute errors about what precision that error is calculated in. A larger precision than what's being tested makes sense, to match the wording from ULP - "the reference value used to compute the ULP value of an arithmetic operation is the infinitely precise result"
which is what this PR does already. If so, I think it is our implementation that should be improved.

TLDR - I think this PR checks dot/cross error correctly, but we could use a spec clarification about the precision those absolute errors are calculated in.

Reply:
Wrote this up as a OpenCL-Doc issue - https://github.com/KhronosGroup/OpenCL-Docs/issues/900